### PR TITLE
clang_9: fix compilation of HIP-code

### PR DESCRIPTION
**Pull request by DieGoldeneEnte**
###### Motivation for this change
When compiling HIP-code clang executes other llvm binaries, such as
llvm-link, llc or opt. Clang expects these binaries in the folder with
the clang-executable (...clang/bin/), this is not true for nix, because
these binaries are in the llvm/bin folder. It is not possible to simply
replace the path of the called library, because then the compilation
with gcc fails. While I personally did not encounter a problem with
clang failing due to calling lld I added it for completeness, this
however means clang also depends on the llvm binutils.

Maybe older clang versions have the same flaw, I did not check this yet since my test setup only works with clang9 at the moment. Also I am not sure whether this is the most elegant solution, since I don't have too much experience with clang or nix.

###### Things done
I checked for successful compilation using [this this setup](https://github.com/DieGoldeneEnte/hipsycl-test). Building this was not possible without this change.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by matthewbauer** [comment]
Does this just effect clang 9 or should we do it for other clang versions as well?

File: pkgs/development/compilers/llvm/9/clang/default.nix [hunk 1]
```diff
@@ -50,6 +50,13 @@ let
              -e 's/Args.hasArg(options::OPT_nostdlibinc)/true/' \
              lib/Driver/ToolChains/*.cpp
 
+      # Patch to force correct executable path for hip (llvm-link,opt,llc,lld)
+      sed -i -e 's|SmallString<128> ExecPath(C.getDriver().Dir);|SmallString<128> ExecPath(C.getDriver().Name.find("clang") != std::string::npos ? "${llvm}/bin" : C.getDriver().Dir);|' \
+             -e 's|SmallString<128> OptPath(C.getDriver().Dir);|SmallString<128> OptPath(C.getDriver().Name.find("clang") != std::string::npos ? "${llvm}/bin" : C.getDriver().Dir);|' \
+             -e 's|SmallString<128> LlcPath(C.getDriver().Dir);|SmallString<128> LlcPath(C.getDriver().Name.find("clang") != std::string::npos ? "${llvm}/bin" : C.getDriver().Dir);|' \
+             -e 's|SmallString<128> LldPath(C.getDriver().Dir);|SmallString<128> LldPath(C.getDriver().Name.find("clang") != std::string::npos ? "${bintools}/bin" : C.getDriver().Dir);|' \
```
Comment (matthewbauer) [inline pkgs/development/compilers/llvm/9/clang/default.nix]: binutils only has lld for clang toolchains. We should use the `lld` package here instead.

**Comment by matthewbauer** [comment]
> however means clang also depends on the llvm binutils

Clang already has a runtime dependency on llvm, so this is no problem (or at the worst just a few extra wrappers).

Ideally we could upstream this to LLVM. I think instead of using `C.getDriver().Dir` here, it would be better to do a find_program in CMake like is done for https://github.com/llvm-mirror/clang/blob/65acf43270ea2894dffa0d0b292b92402f80c8cb/CMakeLists.txt#L91-L92. Perhaps just a ticket for LLVM would do.

**Comment by DieGoldeneEnte** [comment]
As far as I found, the commit introducing the code is present since llvm 7.0.0 (and will currently also be present in llvm 10). I haven't tried versions other than 9, because I haven't found good test code for checking.

**Comment by DieGoldeneEnte** [comment]
I confirmed the problem with clang 8 and 9, for clang7 I would need some more work to get older versions of the device libraries.
I will close this PR while I try to get this fixed upstream, since these changes will not be needed once this is fixed.

**Pull request closed without merging**